File: 프로그래머스/unrated/159993. 미로 탈출/미로 탈출.py
from collections import deque
def solution(maps):
    Y = len(maps)
    X = len(maps[0])
    for y in range(Y):
        for x in range(X):
            if maps[y][x] == 'S':
                sL = [y, x]
            elif maps[y][x] == 'L':
                lL = [y, x]
            elif maps[y][x] == 'E':
                eL = [y, x]
    
    able = False
    Q = deque()
    Q.append(sL)
    V = [[0] * X for _ in range(Y)]
    V[sL[0]][sL[1]] = 1
    lDist = 0
    
    while Q:
        elems = Q.popleft()
        i, j = elems[0], elems[1]
        for ni, nj in [[i-1,j],[i,j+1],[i+1,j],[i,j-1]]:
            if 0 <= ni < Y and 0 <= nj < X and maps[ni][nj] != 'X':
                if not V[ni][nj]:
                    if maps[ni][nj] == 'L':
                        lDist = V[i][j] + 1
                        break
                    else:
                        V[ni][nj] = V[i][j] + 1
                        Q.append([ni,nj])
        if lDist:
            lDist -= 1
            able = True
            break
    
    if able:
        able = False
        Q = deque()
        Q.append(lL)
        V = [[0] * X for _ in range(Y)]
        V[lL[0]][lL[1]] = 1
        eDist = 0
        
        while Q:
            elems = Q.popleft()
            i, j = elems[0], elems[1]
            for ni, nj in [[i-1,j],[i,j+1],[i+1,j],[i,j-1]]:
                if 0 <= ni < Y and 0 <= nj < X and maps[ni][nj] != 'X':
                    if not V[ni][nj]:
                        if maps[ni][nj] == 'E':
                            eDist = V[i][j] + 1
                            break
                        else:
                            V[ni][nj] = V[i][j] + 1
                            Q.append([ni,nj])
            if eDist:
                eDist -= 1
                able = True
                break
    
    if able:
        answer = lDist + eDist
    else:
        answer = -1
    
    return answer

File: 프로그래머스/unrated/159993. 미로 탈출/미로 탈출.js
function solution(maps) {
  const Y = maps.length;
  const X = maps[0].length;
  let sL, lL, eL;

  for (let y = 0; y < Y; y++) {
    for (let x = 0; x < X; x++) {
      if (maps[y][x] === 'S') {
        sL = [y, x];
      } else if (maps[y][x] === 'L') {
        lL = [y, x];
      } else if (maps[y][x] === 'E') {
        eL = [y, x];
      }
    }
  }

  let able = false;
  const Q = [];
  Q.push(sL);
  const V = Array.from({ length: Y }, () => Array(X).fill(0));
  V[sL[0]][sL[1]] = 1;
  let lDist = 0;

  while (Q.length > 0) {
    const elems = Q.shift();
    const i = elems[0];
    const j = elems[1];

    const directions = [[i - 1, j], [i, j + 1], [i + 1, j], [i, j - 1]];

    for (const [ni, nj] of directions) {
      if (0 <= ni && ni < Y && 0 <= nj && nj < X && maps[ni][nj] !== 'X') {
        if (!V[ni][nj]) {
          if (maps[ni][nj] === 'L') {
            lDist = V[i][j] + 1;
            break;
          } else {
            V[ni][nj] = V[i][j] + 1;
            Q.push([ni, nj]);
          }
        }
      }
    }

    if (lDist) {
      lDist -= 1;
      able = true;
      break;
    }
  }
  let eDist = 0;
  if (able) {
    able = false;
    Q.length = 0;
    Q.push(lL);
    const V = Array.from({ length: Y }, () => Array(X).fill(0));
    V[lL[0]][lL[1]] = 1;
    

    while (Q.length > 0) {
      const elems = Q.shift();
      const i = elems[0];
      const j = elems[1];

      const directions = [[i - 1, j], [i, j + 1], [i + 1, j], [i, j - 1]];

      for (const [ni, nj] of directions) {
        if (0 <= ni && ni < Y && 0 <= nj && nj < X && maps[ni][nj] !== 'X') {
          if (!V[ni][nj]) {
            if (maps[ni][nj] === 'E') {
              eDist = V[i][j] + 1;
              break;
            } else {
              V[ni][nj] = V[i][j] + 1;
              Q.push([ni, nj]);
            }
          }
        }
      }

      if (eDist) {
        eDist -= 1;
        able = true;
        break;
      }
    }
  }

  if (able) {
    return lDist + eDist;
  } else {
    return -1;
  }
}
